File: SnowRoll/Assets/Scripts/Libs/UI/UICore/ComponentStyle/WidgetStyleMgr.cs
﻿using System.Collections.Generic;

namespace SDK.Lib
{
    public class WidgetStyleMgr
    {
        protected MDictionary<WidgetStyleID, MDictionary<int, WidgetStyle>> m_style2Dic;

        public WidgetStyleMgr()
        {
            m_style2Dic = new MDictionary<WidgetStyleID, MDictionary<int, WidgetStyle>>();
            registerStype();
        }

        public void addWidgetStype(WidgetStyleID widgetId, int comId, WidgetStyle style)
        {
            if(!m_style2Dic.ContainsKey(widgetId))
            {
                m_style2Dic[widgetId] = new MDictionary<int, WidgetStyle>();
            }
            m_style2Dic[widgetId][comId] = style;
        }

        public T GetWidgetStyle<T>(WidgetStyleID widgetId, int comId) where T : WidgetStyle
        {
            return m_style2Dic[widgetId][comId] as T;
        }

        protected void registerStype()
        {
            LabelStyleBase lblStyle = new LabelStyleBase();
            addWidgetStype(WidgetStyleID.eWSID_Text, (int)LabelStyleID.eLSID_None, lblStyle);

            ButtonStyleBase btnStyle = new ButtonStyleBase();
            addWidgetStype(WidgetStyleID.eWSID_Button, (int)BtnStyleID.eBSID_None, btnStyle);
        }
    }
}

File: Bak/scripts/MyLua/Libs/UI/UICore/ComponentStyle/WidgetStyleMgr.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "WidgetStyleMgr";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mStyle2Dic = GlobalNS.new(GlobalNS.MDictionary);
    self:registerStype();
end

function M:addWidgetStype(widgetId, comId, style)
    if(not self.mStyle2Dic:ContainsKey(widgetId)) then
        local styleDic = GlobalNS.new(GlobalNS.MDictionary);
        self.mStyle2Dic:Add(widgetId, styleDic);
    end
	
    self.mStyle2Dic:value(widgetId):Add(comId, style);
end

function M:GetWidgetStyle(widgetId, comId)
    return self.mStyle2Dic:value(widgetId):value(comId);
end

function M:registerStype()
    local lblStyle = GlobalNS.new(GlobalNS.LabelStyleBase);
    self:addWidgetStype(GlobalNS.WidgetStyleID.eWSID_Text, GlobalNS.LabelStyleID.eLSID_None, lblStyle);

    local btnStyle = GlobalNS.new(GlobalNS.ButtonStyleBase);
    self:addWidgetStype(GlobalNS.WidgetStyleID.eWSID_Button, GlobalNS.BtnStyleID.eBSID_None, btnStyle);
end

return M;